src/Stratis.Bitcoin.Features.BlockStore/BlockStore.cs

```diff
@@ -0,0 +1,392 @@
+            this.dequeueLoopTask = this.DequeueBlocksContinuouslyAsync();
+
+            this.logger.LogTrace("(-)");
+        }
+        
+        /// <summary>Sets the internal store tip and exposes the store tip to other components through the chain state.</summary>
+        private void SetStoreTip(ChainedHeader newTip)
+        {
+            this.storeTip = newTip;
+            this.chainState.BlockStoreTip = newTip;
+        }
+
+        /// <summary>
+        /// Sets block store tip to the last block that exists both in the repository and in the <see cref="ConcurrentChain"/>.
+        /// </summary>
+        private async Task RecoverStoreTipAsync()
+        {
+            this.logger.LogTrace("()");
+
+            var blockStoreResetList = new List<uint256>();
+            uint256 resetBlockHash = this.blockRepository.BlockHash;
+            Block resetBlock = await this.blockRepository.GetAsync(resetBlockHash).ConfigureAwait(false);
+
+            while (this.chain.GetBlock(resetBlockHash) == null)
+            {
+                blockStoreResetList.Add(resetBlockHash);
+
+                if (resetBlock.Header.HashPrevBlock == this.chain.Genesis.HashBlock)
+                {
+                    resetBlockHash = this.chain.Genesis.HashBlock;
+                    break;
+                }
+
+                resetBlock = await this.blockRepository.GetAsync(resetBlock.Header.HashPrevBlock).ConfigureAwait(false);
+
+                if (resetBlock == null)
+                {
+                    // This can happen only if block store is corrupted.
+                    throw new BlockStoreException("Block store failed to recover.");
+                }
+
+                resetBlockHash = resetBlock.GetHash();
+            }
+
+            ChainedHeader newTip = this.chain.GetBlock(resetBlockHash);
+
+            if (blockStoreResetList.Count != 0)
+                await this.blockRepository.DeleteAsync(newTip.HashBlock, blockStoreResetList).ConfigureAwait(false);
+
+            this.SetStoreTip(newTip);
+
+            this.logger.LogWarning("Block store tip recovered to block '{0}'.", newTip);
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// Adds a block to the saving queue.
+        /// </summary>
+        /// <param name="blockPair">The block and its chained header pair to be added to pending storage.</param>
+        public void AddToPending(BlockPair blockPair)
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(blockPair), blockPair.ChainedHeader);
+
+            this.blocksQueue.Enqueue(blockPair);
+
+            this.logger.LogTrace("(-)");
+        }
+        
+        /// <summary>
+        /// Dequeues the blocks continuously and saves them to the database when max batch size is reached or timer ran out.
+        /// </summary>
+        /// <remarks>Batch is always saved on shutdown.</remarks>
+        private async Task DequeueBlocksContinuouslyAsync()
```

Comment:
I still feel that this method does too much and should be broken up into separate methods to make it a bit more readable. This whole method can be rewritten like so (method names can be improved):

@dangershony @bokobza @majikandy  Thoughts?

```
        private async Task DequeueBlocksContinuouslyAsync()
        {
            this.logger.LogTrace("()");

            var batch = new List<BlockPair>();

            Task<BlockPair> dequeueTask = null;
            Task timerTask = null;

            while (!this.nodeLifetime.ApplicationStopping.IsCancellationRequested)
            {
                var saveBatch = await DequeueBlockAsync(dequeueTask, timerTask);
                ProcessDequeuedTask(dequeueTask, saveBatch, batch);
                SaveBatch(saveBatch, batch, timerTask);
            }

            this.logger.LogTrace("(-)");
        }
```

Reply:
Probably the only good way to improve on it as @Aprogiena suggested is to introduce an AsyncQueue with timer that will have that logic (dequeuing on timer or condition).

I can do that in the next PR if you'd like (introduce AQ with timer and use it here and in BS signaled)

Reply:
>The problem with that method is that if you are to split it into several small methods you will need to pass almost every variable as a ref

Won't they all just be by ref anyway?  these aren't primitives are they you are talking about passing around?

>So in order to simplify method which has 20 lines that do something we are going to have an additional class just for this method and 1-3 more methods? Sounds like making it more complicated.

Additional classes for readable code, testable code, single responsibility principle are often a valuable change over 20 lines of logic here, then 20 more further down, then 20 more over there when it comes to maintainable code.  Files are cheap, classes are cheap. Use them.

Reply:
Main point is- just moving the logic to different methods wont make it simpler. 

Moving the dequeuing logic to a new queue with timer, however, will.

Reply:
>This approach was copied from the BlockStoreSignaled almost 1 to 1.  

yes and I did not like that either 😉 

>just moving the logic to different methods wont make it simpler.

I don't necessarily agree, moving code around will make the main flow more readable, currently to understand the flow I need to read `timertask = (timertask != null) ? Task.WhenAn(.....`
This could be replaced with a `GetTimerTaskOrDequeuTask` which could be argued is more understandable. 

Hmm so apro propose an AsyncQueue that will have a timer internally?
Not a bad idea, but it should be a new class (that utilizes AsyncQueue and not build it in to the existing class perhaps an `AsyncQueueScheduler`)

Please dont make any changes yet lets talk tomorrow

Reply:
>Hmm so apro propose an AsyncQueue that will have a timer internally?

Yes. Basically the same idea as we once had when we were working on bsSignaled. Except for that time timer wasn't added because it was the only place where such logic was needed. 
Now we have 2 places (and potentially much more)

>Please dont make any changes yet lets talk tomorrow

Sure. Adding an AsynqQueueWithTimer is a huge task anyway. If we are to do this- for sure it should be done outside of this PR.

Reply:
@fassadlr it was agreed that at a later stage we will introduce a new `AsyncTimerQueue` that will greatly reduce the complexity/size of this method.